encoder_clip.py

```diff
@@ -0,0 +1,91 @@
+from typing import Optional, Iterable, Any, List
+
+import numpy as np
+import torch
+from PIL import Image
+from jina import Executor, DocumentArray, requests
+
+import clip
+
+
+def _batch_generator(data: List[Any], batch_size: int):
+    for i in range(0, len(data), batch_size):
+        yield data[i:min(i + batch_size, len(data))]
+
+
+class CLIPImageEncoder(Executor):
+    """
+    Encode image into embeddings.
+
+    :param model_name: use clip.available_models() to see all available models: ['RN50', 'RN101', 'RN50x4', 'ViT-B/32']
+    :param device: device to use for encoding ['cuda', 'cpu] - if not set, the device is detected automatically
+    :param default_batch_size: fallback batch size in case there is not batch size sent in the request
+    :param default_traversal_path: fallback traversal path in case there is not traversal path sent in the request
+    :param jit: Whether to load the optimized JIT model (default) or more hackable non-JIT model.
+    :param channel_axis: axis of the color channel
+    """
+
+    def __init__(
+            self,
+            model_name: str = 'ViT-B/32',
+            use_default_preprocessing: bool = True,
+            device: Optional[str] = None,
+            default_batch_size: int = 32,
+            default_traversal_path: str = 'r',
+            jit: bool = True,
+            *args, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        if not device:
+            device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        self.device = device
+        self.default_batch_size = default_batch_size
+        self.default_traversal_path = default_traversal_path
+        model, preprocess = clip.load(model_name, device, jit)
+        self.model = model
+        self.use_default_preprocessing = use_default_preprocessing
+        self.preprocess = preprocess
+
+    @requests
+    def encode(self, docs: Optional[DocumentArray], parameters: dict, **kwargs):
```

Comment:
why is docs optional and not parameters? is it tested without sending them?

Reply:
parameters have an empty dictionary as default value and won't be None.
But if you don't insert docs, instead of retrieving an empty DocumentArray it can be None.
I added a test.